### PR TITLE
Make feign.Request.Options as refreshable beans inside Feign clients

### PR DESCRIPTION
**Pull request by jas34**
Refer: #468

- Created OptionsFactoryBean to create instance of Options bean with applicable configuration properties.
- FeignClientRegistrar:
    - created a private method to populate Bean definition for Options bean.
    - invoked private method during Feign bean definition registration phase.
- FeignClientFactoryBean:
    - updated to get refreshable options bean from feign context by bean name.
- FeignContext:
    - exposed a method to get bean from context by bean name.
- added test cases.

Fixes gh-468.

### CODE REVIEW

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Hello, @jas34 . Thanks for submitting the PR. I have added some comments.
Also, I have run some tests and have verified that the `Options` properties do not seem to be properly passed on to the request when `refresh-enabled` is set to `true` - please verify.
In order to reproduce, use [this sample](https://github.com/spring-cloud-samples/feign-eureka), modify the SC version to `2020.0.3-SNAPSHOT` to pick your locally built changes and update [`application.yml`](https://gist.github.com/OlgaMaciaszek/ca501419887c588edf97292138fefe7e). Then when you debug at [`Client#execute`]( https://github.com/OpenFeign/feign/blob/3ea3708cddef0d20c0bf443c18f2364c8d8df062/core/src/main/java/feign/Client.java#L104) the value set in properties is not passed on correctly, however with `refresh-enabled` set to `false` it is.

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientFactoryBean.java [hunk 1]
```diff
@@ -486,6 +511,10 @@ public void setFallbackFactory(Class<?> fallbackFactory) {
 		this.fallbackFactory = fallbackFactory;
 	}
 
+	void setRefreshableClient(boolean refreshableClient) {
+		this.refreshableClient = refreshableClient;
+	}
+
 	@Override
```
[the remark this answers is not in the document]
Comment (jas34) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientFactoryBean.java:517]: respective methods updated

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientsRegistrar.java [hunk 2]
```diff
@@ -408,4 +415,28 @@ public void setEnvironment(Environment environment) {
...
+			definitionBuilder.addPropertyValue("contextId", contextId);
+			BeanDefinitionHolder definitionHolder = new BeanDefinitionHolder(definitionBuilder.getBeanDefinition(),
+					beanName);
+			definitionHolder = ScopedProxyUtils.createScopedProxy(definitionHolder, registry, true);
+			BeanDefinitionReaderUtils.registerBeanDefinition(definitionHolder, registry);
+		}
+	}
+
+	private boolean isClientRefreshEnabled() {
+		return environment.getProperty("feign.client.refresh-enabled", Boolean.class, false);
```
[the remark this answers is not in the document]
Comment (jas34) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientsRegistrar.java:439]: property added to `additional-spring-configuration-metadata.json`

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientFactoryBean.java [hunk 3]
```diff
@@ -334,6 +344,21 @@ protected void configureUsingProperties(FeignClientProperties.FeignClientConfigu
 				"No Feign Client for loadBalancing defined. Did you forget to include spring-cloud-starter-loadbalancer?");
 	}
 
+	/**
+	 * meant to get Options bean from context with bean name.
```
[the remark this answers is not in the document]
Comment (jas34) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientFactoryBean.java]: Updated.

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientFactoryBean.java [hunk 4]
```diff
@@ -334,6 +344,21 @@ protected void configureUsingProperties(FeignClientProperties.FeignClientConfigu
...
+	 * @param context context of Feign client
+	 * @param contextId name of feign client
+	 * @return
+	 */
+	protected Request.Options getOptionsByName(FeignContext context, String contextId) {
+		if (refreshableClient) {
+			return context.getInstance(contextId, Request.Options.class.getCanonicalName() + "-" + contextId,
+					Request.Options.class);
+		}
+
```
[the remark this answers is not in the document]
Comment (jas34) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientFactoryBean.java]: removed.

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientFactoryBean.java [hunk 5]
```diff
@@ -486,6 +511,10 @@ public void setFallbackFactory(Class<?> fallbackFactory) {
 		this.fallbackFactory = fallbackFactory;
 	}
 
+	void setRefreshableClient(boolean refreshableClient) {
```
[the remark this answers is not in the document]
Comment (jas34) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignClientFactoryBean.java]: updated

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientWithRefreshableOptionsTest.java [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+	private Application.ConnectTimeoutClient connectTimeoutClient;
+
+	@Autowired
+	private Application.OverrideOptionsClient overrideOptionsClient;
+
+	@Autowired
+	private FeignClientProperties clientProperties;
+
+	@Test
+	@DisabledForJreRange(min = JRE.JAVA_16)
```
[the remark this answers is not in the document]
Comment (jas34) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientWithRefreshableOptionsTest.java]: Done

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientWithRefreshableOptionsTest.java [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+	public void refreshScopeBeanDefinitionShouldBePresent() {
+		BeanDefinition beanDefinition = ((GenericWebApplicationContext) applicationContext)
+				.getBeanDefinition(Request.Options.class.getCanonicalName() + "-" + "refreshableClient");
+		BeanDefinition originBeanDefinition = beanDefinition.getOriginatingBeanDefinition();
+		assertThat(originBeanDefinition.getBeanClassName()).isEqualTo(OptionsFactoryBean.class.getCanonicalName());
+		assertThat(originBeanDefinition.getScope()).isEqualTo("refresh");
+	}
+
+	@Test
+	@DisabledForJreRange(min = JRE.JAVA_16)
```
[the remark this answers is not in the document]
Comment (jas34) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientWithRefreshableOptionsTest.java]: Done

**Comment by jas34** [comment]
> Hello, @jas34 . Thanks for submitting the PR. I have added some comments.
> Also, I have run some tests and have verified that the `Options` properties do not seem to be properly passed on to the request when `refresh-enabled` is set to `true` - please verify.
> In order to reproduce, use [this sample](https://github.com/spring-cloud-samples/feign-eureka), modify the SC version to `2020.0.3-SNAPSHOT` to pick your locally built changes and update [`application.yml`](https://gist.github.com/OlgaMaciaszek/ca501419887c588edf97292138fefe7e). Then when you debug at [`Client#execute`](https://github.com/OpenFeign/feign/blob/3ea3708cddef0d20c0bf443c18f2364c8d8df062/core/src/main/java/feign/Client.java#L104) the value set in properties is not passed on correctly, however with `refresh-enabled` set to `false` it is.

@OlgaMaciaszek I have checked as per the [sample ](https://github.com/spring-cloud-samples/feign-eureka) application provided.
I believe the proxy snapshot shown by the debugger has been misleading for you. When `feign.client.refresh-enabled`=`true` a scoped target CGLIB proxy bean gets created which does not show correct value in debugger because actual values are fetched when ObjectFactory.getObject() (in this case `org.springframework.cloud.openfeign.OptionsFactoryBean`) is invoked by FactoryBeanRegistrySupport.
In this case when options.connectTimeoutMillis() or any method of options beans is invoked it will return the correct value like 200 it is returning right now.
I have created a sample [POC](https://github.com/jas34/feign-eureka/blob/526-testing/client/src/main/java/demo/HelloClientApplication.java) (Refreshscope TestBean) to verify the same. There also the default value of field value will be 0 in debugger but returns an actual value of 10.

Let me know if you still find any issues.

**Comment by OlgaMaciaszek** [comment]
>@OlgaMaciaszek I have checked as per the sample application provided.
I believe the proxy snapshot shown by the debugger has been misleading for you. When feign.client.refresh-enabled=true a scoped target CGLIB proxy bean gets created which does not show correct value in debugger because actual values are fetched when ObjectFactory.getObject() (in this case org.springframework.cloud.openfeign.OptionsFactoryBean) is invoked by FactoryBeanRegistrySupport.
In this case when options.connectTimeoutMillis() or any method of options beans is invoked it will return the correct value like 200 it is returning right now.
I have created a sample POC (Refreshscope TestBean) to verify the same. There also the default value of field value will be 0 in debugger but returns an actual value of 10.

Let me know if you still find any issues.
Ok, I have checked with some logging and in fact the values passed via debugger were incorrect. Will run some more tests.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientWithRefreshableOptionsTest.java [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+	}
+
+	@Test
+	public void readTimeoutShouldWorkWhenConnectTimeoutNotSet() {
+		Request.Options options = getRequestOptions((Proxy) readTimeoutClient);
+		assertOptions(options, 5000, 2000);
+	}
+
+	@Test
+	public void connectTimeoutShouldWorkWhenReadTimeoutNotSet() {
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientWithRefreshableOptionsTest.java:104]: I did not mean to just remove the annotations but to only add tests that will run correctly also on JDK 16. These tests don't. Please modify the tests.

Comment (jas34) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientWithRefreshableOptionsTest.java:104]: I will check and get back to you. I thought it has been made compatible

Comment (jas34) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientWithRefreshableOptionsTest.java:97]: I have updated test strategy.
Implemented `feign.Client` with `OptionsTestClient` to return applicable Options from client as response `OptionsResponseForTests` and then use it for assertions.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
@jas34 The added code looks good now. Please merge current `master` changes and resolve conflicts and document the newly added feature (in `spring-cloud-openfeign.adoc`), then run build with -Pdocs` before pushing to recreate auto-generated files.

File: docs/src/main/asciidoc/spring-cloud-openfeign.adoc [hunk 9]
```diff
@@ -717,6 +717,17 @@ You may consider enabling Jackson Modules for the support `org.springframework.d
 feign.autoconfiguration.jackson.enabled=true
 ----
 
+=== Spring RefreshScope Support
```
Comment (OlgaMaciaszek) [inline docs/src/main/asciidoc/spring-cloud-openfeign.adoc:720]: Change to Spring `@RefreshSchope` Support

File: docs/src/main/asciidoc/spring-cloud-openfeign.adoc [hunk 10]
```diff
@@ -717,6 +717,17 @@ You may consider enabling Jackson Modules for the support `org.springframework.d
...
+If feign client refresh is enabled then each feign client is created with `feign.Request.Options` as a refresh scoped bean. This means properties like `connectTimeout` and `readTimeout` can be refreshed against any feign client instance via `POST /actuator/refresh`.
```
Comment (OlgaMaciaszek) [inline docs/src/main/asciidoc/spring-cloud-openfeign.adoc:721]: Pleasse capitalise "feign". Please add a hyphen in between in "refresh scoped".

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
@jas34 Have added some small comments regarding the documentation. Also, I am requesting @Buzzardo, to review the documentation changes.

File: docs/src/main/asciidoc/spring-cloud-openfeign.adoc (hunk 10 above)
Comment (jas34) [reply docs/src/main/asciidoc/spring-cloud-openfeign.adoc]: updates done

Comment (Buzzardo) [reply docs/src/main/asciidoc/spring-cloud-openfeign.adoc]: Change to (incorporate Olga's request and my request):

If Feign client refresh is enabled, each feign client is created with `feign.Request.Options` as a refresh-scoped bean. This means properties such as `connectTimeout` and `readTimeout` can be refreshed against any Feign client instance through `POST /actuator/refresh`.

File: docs/src/main/asciidoc/spring-cloud-openfeign.adoc [hunk 11]
```diff
@@ -717,6 +717,17 @@ You may consider enabling Jackson Modules for the support `org.springframework.d
...
+
+By default refresh behaviour in feign clients is disabled. Use the following property to enable refresh behaviour.
```
Comment (Buzzardo) [inline docs/src/main/asciidoc/spring-cloud-openfeign.adoc]: Change to:

By default, refresh behavior in Feign clients is disabled. Use the following property to enable refresh behavior:

File: docs/src/main/asciidoc/spring-cloud-openfeign.adoc [hunk 12]
```diff
@@ -717,6 +717,17 @@ You may consider enabling Jackson Modules for the support `org.springframework.d
...
+[source,java]
+----
+feign.client.refresh-enabled=true
+----
+TIP: DO NOT annotate `@FeignClient` interface with `@RefreshScope` annotation.
```
Comment (Buzzardo) [inline docs/src/main/asciidoc/spring-cloud-openfeign.adoc:728]: Change to:

TIP: DO NOT annotate the `@FeignClient` interface with the `@RefreshScope` annotation.

**Review by Buzzardo — CHANGES_REQUESTED** [review]

File: docs/src/main/asciidoc/spring-cloud-openfeign.adoc (hunk 11 above)
Comment (jas34) [reply docs/src/main/asciidoc/spring-cloud-openfeign.adoc]: changed.

File: docs/src/main/asciidoc/spring-cloud-openfeign.adoc (hunk 12 above)
Comment (jas34) [reply docs/src/main/asciidoc/spring-cloud-openfeign.adoc]: changed

**Review by Buzzardo — APPROVED** [review]

File: docs/src/main/asciidoc/spring-cloud-openfeign.adoc (hunk 9 above)
Comment (OlgaMaciaszek) [reply docs/src/main/asciidoc/spring-cloud-openfeign.adoc:720]: @jas34 This one change is still missing.

Comment (OlgaMaciaszek) [reply docs/src/main/asciidoc/spring-cloud-openfeign.adoc:720]: Please add "@" and surround "@RefreshScope" with "`"

Comment (jas34) [reply docs/src/main/asciidoc/spring-cloud-openfeign.adoc]: Oh! I missed this.
Added as suggested above

**Review by OlgaMaciaszek — APPROVED** [review]
LGTM.

**Pull request merged**